dask/core.py

```diff
@@ -224,27 +186,34 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
     set(['x'])
     """
     if key is not None:
-        args = [dsk[key]]
+        arg = dsk[key]
     elif task is not None:
-        args = [task]
+        arg = task
     else:
         raise ValueError("Provide either key or task")
+
     result = []
-    while args:
-        arg = args.pop()
-        typ = type(arg)
-        if typ is tuple and arg and callable(arg[0]):  # istask(arg):
-            args.extend(arg[1:])
-        elif type(arg) is list:
-            args.extend(arg)
-        else:
-            result.append(arg)
-    if not result:
-        return [] if as_list else set()
-    rv = []
-    for x in result:
-        rv.extend(_deps(dsk, x))
-    return rv if as_list else set(rv)
+    work = [arg]
+
+    while work:
+        new_work = []
+        for w in work:
+            typ = type(w)
+            if typ is tuple and w and callable(w[0]):  # istask(w)
+                new_work += w[1:]
+            elif typ is list:
+                new_work += w
+            elif typ is dict:
```

Comment:
Supporting dict values seems like a natural extension to me.  More natural than the `(dict, ['a'], ['x'])` solution.